### ISSUE TITLE
Ordering and space membership

### ISSUE THREAD

**Issue by kohr-h** [odlgroup/odl#964]
I'm struggling with the `order` property and how to treat space equality and membership of elements.

In some sense, `'A'` is more general than `'C'` and `'F'` ordering since a C- or F-contiguous tensor is also a valid member of an A-ordered space. So what I would find reasonable (or rather the opposite would be weird) is the following:
```python
>>> space_c = odl.rn((2, 3), order='C')
>>> space_a = odl.rn((2, 3), order='A')  # default
>>> space_a.element() in space_c
False
>>> space_c.element() in space_a
True
```
There is a related issue related to slicing into tensors. If the original tensor is in a `'C'` space, what should a non-contiguously sliced thing be? Should it be made contiguous in the original space's order or should it drop the contiguousness guarantee? Currently non-contiguous slices result in `'A'` spaces, which saves from creating a new array, but also has this effect:
```python
>>> space = odl.rn((2, 3, 4), order='C')
>>> x = space.element()
>>> x[..., 0].order
'A'
>>> sliced_space = space[:-1]
>>> x[..., 0] in sliced_space
False
>>> # Wat??
>>> sliced_space.order
'C'
```
After writing this up, I actually only see one reasonable solution, which is to force `x[..., 0]` to also be C contiguous. That makes slicing inefficient in that case, so the F and C spaces would be much less useful.

Then  we also have the interesting case of spaces with `ndim=1`, where C and F ordering are equivalent. What do we want there?
```python
>>> space_c = odl.rn(3, order='C')
>>> space_a = odl.rn(3, order='A')
>>> space_f = odl.rn(3, order='F')
>>> space_c.element() in space_f  # reasonable?
True
>>> space_f.element() in space_c  # reasonable?
True
>>> space_a.element() in space_c
False
```
Is this something sane? We do want to keep the spaces apart since we can do this
```python
>>> space_a[[0, 0]]
rn((3, 3))
>>> space_c[[0, 0]]
rn((3, 3), order='C')
```
where there would be an actual difference.

### Radical solution
The radical thing would be to simply throw out the `order` property and always use what's now `'A'`. Then we would also have less pain adding a `strides` option later.
Users who want contiguous arrays (e.g. for interfacing with some external library) should call their method of choice explicitly to make a tensor contiguous. Or we keep the `order` option in `asarray` to provide that functionality.

**Comment by adler-j** [comment]
Your radical solution would totally break any hope of no-overhead compatibility with libraries such as GPUMCI and ASTRA, since they have required storage orders that we would need to follow. So in my opinion, such ideas are not acceptable in the long term.

With that said, we could perhaps scrap the option for now, and later bring it back in a more complete version (specifically with explicit strides).

> If the original tensor is in a 'C' space, what should a non-contiguously sliced thing be?

Frankly, to me this is a minor issue, the "waat" example you gave, at least to me, makes perfect sense. And any user advanced enough to care about the single copy that using `order='C'` will save should also be able to figure this out.

My suggestion would be:

Always strive for the most performance-optimal solution unless it is downright stupid. In this case, return a view in all cases, if that causes the order of the space to change, then let it be so. Users can, without cost, later cast the space into the order they want but if we give them the wrong space they are screwed.

**Comment by kohr-h** [comment]
> Your radical solution would totally break any hope of no-overhead compatibility with libraries such as GPUMCI and ASTRA, since they have required storage orders that we would need to follow. So in my opinion, such ideas are not acceptable in the long term.

The thing is that internally we should use views as much as possible, and for that it is most natural not to have an `order` property *on the space* since it's hard to keep promises. IMO it would make more sense to have `order` as optional parameter in the methods `element` (mandatory if used with `data_ptr`) and `asarray` that talk to the outside world.

I also don't quite see where this is going to slow down things significantly. The `np.ascontiguousarray` and `np.asfortranarray` calls will be no-ops in case the array already is in that order, if I'm not totally off.

If we do this:
```python
# In TensorSpace
def element(self, inp, order='A'):
    ...
    if inp in self and order == 'A':
        # Short-circuit for space elements
        return inp

    arr = np.array(inp, copy=False, dtype=self.dtype, ndmin=self.ndim,
                   order=order)
    ...
```
and this:
```python
# In Tensor
def asarray(self, out=None, order='A'):
    if out is None:
        if order == 'C':
            return np.ascontiguousarray(self.data)
        elif order == 'F':
            return np.asfortranarray(self.data)
        else:
            return self.data
    else:
        out[:] = self.data
        return out
```
where would the overhead be?

**Comment by adler-j** [comment]
Well consider this use case:

```
def call_my_weird_library(x):
    # library requires that x has 'F' ordering. Also note that x is stored on a GPU in norther china so copies are *expensive*
    # The result is written in place
```

Any method that calls this would ofc do something as you say

```
x_f = np.asfortranarray(x)
call_my_weird_library(x_f)
x[:] = x_f
```

But unless we have a way of guaranteeing that `x` is 'F' ordered, this will simply be a waste of time and space, and we would be forced to assume that two copies will happen (and twice the memory used).

Simple stuff like

```
1 + vec
```

would break the ordering, not to mention that any operator that does not have a non-out-parameter version calls `space.element` under the hood, and we have absolutely no way to change how that is called.

**Comment by kohr-h** [comment]
> Well consider this use case:
> 
> def call_my_weird_library(x):
>     # library requires that x has 'F' ordering. Also note that x is stored on a GPU in norther china so copies are *expensive*
>     # The result is written in place
> 
> Any method that calls this would ofc do something as you say
> 
> x_f = np.asfortranarray(x)
> call_my_weird_library(x_f)
> x[:] = x_f
> 
> But unless we have a way of guaranteeing that x is 'F' ordered, this will simply be a waste of time and space, and we would be forced to assume that two copies will happen (and twice the memory used).

So let's make this more precise. I guess in your library in China expects some object it knows. What does it do with an ODL vector? No idea. 
Probably what it *anyway* has to do is to convert to a Numpy array, and there's your expensive transfer. *Or* the ODL space that implements this special kind of representation has a method that returns the underlying Chinese array.
So the weird library expects some kind of data representation, and the ODL space needs to have a way to provide it entirely locally, without transmitting anything to the other side of the Great Wall. Some kind of `asarray` method that does this locally, if provided with `order` even guaranteed to be in specific ordering.
For example
```python
>>> executor = ChinaLibExecutor()
>>> array_in_china = odl.rn(int(1e8), impl='china').element()
>>> future = exectuor.submit(np.asfortranarray, array_in_china)
```
All good, runs remotely.

So I don't know what you were getting at, but if you run something that expects a Fortran-ordered *Numpy* array, it's gonna copy. If you want another type of array, you need custom methods. And a method checking if you have F contiguousness is simple for Numpy, and for the weird library, you're going to have something similar.

> Simple stuff like
> 
> 1 + vec
> 
> would break the ordering, not to mention that any operator that does not have a non-out-parameter version calls space.element under the hood, and we have absolutely no way to change how that is called.

That's a point. But if you try to keep a fixed ordering at every time, you can do simple stuff like `1 + vec`, but anything that you do with slicing or summing along axes, or other stuff that break contiguousness, you get a copy every time. So what is worse? Copying once when interfacing, or copying a couple of times in between?

That issue doesn't go away either when we implement `strides` instead of `order`. Operators never know what to do.

Okay, what to do? I'll think about this some more, but it's too early to remove `order` as it seems.

**Comment by adler-j** [comment]
Well, to me adding `strides` would be the best solution and in that case these questions in general have a much more obvious answer (i.e. slicing will ofc break the strides).

Anyway, I feel the problem here is that you are very numpy specific, and this problem is largely exaggerated when you use other storage types (like GPUs). I'll try to come up with a more specific example, but to me the archetypical example would be a tightly coupled range for a ASTRA based `RayTransform`, where selecting the strides properly would save us an expensive copy.

**Comment by kohr-h** [comment]
Closed by #1088
